### ISSUE TITLE
Integration of __torch_dispatch__ with masked tensor semantics

### ISSUE THREAD

**Issue by anjali411** [pytorch/pytorch#65619]
Creating this issue to discuss if masked tensor semantics are correctly maintained with __torch_dispatch__. This came up in an offline discussion with @ezyang while talking about `DiagLinearOperator` class and the requirement to have gradients of the same size as that of the tensor stored in `DiagLinearOperator` (and not the meta tensor) while using __torch_dispatch__. Currently, the masked semantics for `DiagLinearOperator` are not maintained in the backward pass (the size of gradient tensor is the same as that of the meta tensor).

**Comment by albanD** [comment]
I think that this is not torch_dispatch or the autograd engine's role to enforce this.
In particular, there are cases where we want such a behavior (embedding layers generating sparse gradients for Dense Tensors). So there isn't a strict rule we want to enforce for one type to match the other.

You can actually see that in the masked Tensor case where you not only want the grad to be a masked Tensor but also have the same mask.

I think only the user that implements the subclass knows what is acceptable or not here and should ensure that their implementation returns the right type of output.

**Comment by anjali411** [comment]
right but are there cases when "generating dense gradients for sparse Tensors" is a desirable outcome?

**Comment by albanD** [comment]
For sparse Tensor (as in only storing 0s in a more effective way), then it is always the desirable outcome! As 0 value does not imply 0 gradients.

**Comment by ezyang** [comment]
We discussed this a bit in the composability sync last Friday. Some notes:

* In general, the derivative formulas in PyTorch core should be thought of as being written solely to do dense semantics. So if you wanted to have AD work with masked semantics operators, you're going to have to redefine AD for all of them.
* That being said, you don't technically need to redefine all of the derivative formulas; you can always recover the sparsity pattern by treating the diagonal tensor as a reparametrization: e.g., `1D tensor -(diagonal embed into)-> 2D square matrix -(do matrix multiply)-> ...`. Then it doesn't matter that the derivative pattern for matrix multiply will result in a dense gradient, as the backwards for the parametrization will apply the appropriate masking. This is bad for performance--you want to fuse the diagonal parametrization with the matrix multiply (both in forwards and backwards)--but at least the semantics are fine. For performance, you're still on the hook (without some sort of compiler) for defining backwards for each of these fused reparametrize-and-op composites. Seems like a good use case for compiler.

cc @Balandat

**Comment by bertmaher** [comment]
I'm intrigued by the possibility of applying a compiler here (and I think it's entirely possible) but I'm not super familiar with masked tensors (or, frankly, autograd) so I want to understand exactly what's desired here.

We are considering a case like this:
```
x = torch.randn(3, 3)
y = torch.diag(torch.tensor([3.0, 4.0, 5.0])).requires_grad_()
z = x @ y
z.sum().backward()
"""
y.grad = tensor([[0.3454, 0.3454, 0.3454],
        [2.4759, 2.4759, 2.4759],
        [1.4637, 1.4637, 1.4637]])
"""
```

There are two optimizations we'd like to have here.  (1) `z = x @ y` should be a sparse matmul b/c `y` is sparse, and (2) `y.grad` should be sparse?  (Is y.grad sparse?)

**Comment by albanD** [comment]
1) Yes, the idea here is that this would be done by the user subclass.
2) no the gradient for y is dense as expected. You need to set `requires_grad_()` on the 1D Tensor if you want only a 1D gradient.

**Comment by bertmaher** [comment]
Aha, thanks @albanD .  So to correct my example:
```
>>> x = torch.randn(3, 3)
>>> y = torch.tensor([3.0, 4.0, 5.0], requires_grad=True)
>>> z = x @ torch.diag(y)
>>> z.sum().backward()
>>> y.grad
tensor([-1.7600, -0.9231,  0.0743])
```

**Comment by ezyang** [comment]
Actually seeing the code with `unsqueeze` in it made me realize that there's actually a very generic, non linear operator specific optimization that applies here, which is any time you have "expanding view op -> operation", in the backwards, if you can fuse "backward operation + narrowing view op" this should almost always be profitable.

**Comment by Balandat** [comment]
To provide some follow-up context, we had a discussion on how things should work for structured tensors consisting of lower-dim / structured representations (e.g. the diagonal mentioned above, or a `LowRankTensor` represented internally as `t1 @ t2`. In that case we don't actually want the gradients to be dense and instead have autograd compute gradients of the constituent tensors. So in that case we wouldn't be able to use `__torch_dispatch__`. 

We are considering whether it would make sense to use `__torch_dispatch__` on structured tensors that have a dense representation (e.g. a positive semidefinite tensor) for which we may want to allow gradients themselves to be structured tensors as well. 

cc @Lezcano, @nikitaved, @gpleiss, @jacobrgardner, @j-wilson, @JonathanWenger
